File: erlang/ch1/ex107.erl
%%%-------------------------------------------------------------------
%%% @author Yohan
%%% @copyright (C) 2018, <COMPANY>
%%% @doc
%%%
%%% @end
%%% Created : 27. Jul 2018 2:36 PM
%%%-------------------------------------------------------------------
-module(ex107).
-author("Yohan").

%% API
-export([sqrt/1]).


square(X) -> X * X.

average(X, Y) -> (X + Y) / 2.0.

improve(Guess, X) -> average(Guess, X / Guess).

good_enough(Guess, X) -> abs(square(Guess) - X) < (Guess / 100000.0).

sqrt_iter(Guess, X) ->
  case good_enough(Guess, X) of
    true -> Guess;
    false -> sqrt_iter(improve(Guess, X), X)
  end.

sqrt(X) -> sqrt_iter(1.0, X).

File: julia/ch1/ex107.jl
#=
ex107:
- Julia version: 0.6.4
- Author: Yohan
- Date: 2018-08-01
=#

function square(x)
  x * x
end

function average(x, y)
  (x + y) / 2.0
end

function improve(guess, x)
  average(guess, x / guess)
end

function good_enough(guess, x)
  abs(square(guess) - x) < (guess / 100000.0)
end

function sqrt_iter(guess, x)
  good_enough(guess, x) ? guess : sqrt_iter(improve(guess, x), x)
end

function my_sqrt(x)
  sqrt_iter(1.0, x)
end
